Libraries/CustomComponents/ListView/ListViewDataSource.js

```diff
@@ -112,31 +125,57 @@ class ListViewDataSource {
    * The constructor takes in a params argument that can contain any of the
    * following:
    *
+   * - rows
+   * - rowIdentities
+   * - sectionIdentities
    * - getRowData(dataBlob, sectionID, rowID);
    * - getSectionHeaderData(dataBlob, sectionID);
    * - rowHasChanged(prevRowData, nextRowData);
    * - sectionHeaderHasChanged(prevSectionData, nextSectionData);
    */
-  constructor(params: ParamType) {
-    invariant(
-      params && typeof params.rowHasChanged === 'function',
-      'Must provide a rowHasChanged function.'
-    );
-    this._rowHasChanged = params.rowHasChanged;
-    this._getRowData = params.getRowData || defaultGetRowData;
-    this._sectionHeaderHasChanged = params.sectionHeaderHasChanged;
-    this._getSectionHeaderData =
-      params.getSectionHeaderData || defaultGetSectionHeaderData;
+  constructor({rows, sectionIdentities, rowIdentities, rowHasChanged, getRowData,
+              sectionHeaderHasChanged, getSectionHeaderData}: ParamType) {
+    this._rowHasChanged = rowHasChanged || defaultRowHasChanged;
+    this._getRowData = getRowData || defaultGetRowData;
+    this._getSectionHeaderData = getSectionHeaderData || defaultGetSectionHeaderData;
+    this._sectionHeaderHasChanged = sectionHeaderHasChanged;
+
+    if (rows) {
+      if (rows.constructor === Array) {
+        this._dataBlob = {s1: rows};
+      } else {
+        this._dataBlob = rows;
+      }
 
-    this._dataBlob = null;
-    this._dirtyRows = [];
-    this._dirtySections = [];
-    this._cachedRowCount = 0;
+      if (sectionIdentities) {
+        this.sectionIdentities = sectionIdentities;
+      } else {
+        this.sectionIdentities = Object.keys(this._dataBlob);
+      }
 
-    // These two private variables are accessed by outsiders because ListView
-    // uses them to iterate over the data in this class.
-    this.rowIdentities = [];
-    this.sectionIdentities = [];
+      if (rowIdentities) {
+        this.rowIdentities = rowIdentities;
+      } else {
+        this.rowIdentities = [];
+        this.sectionIdentities.forEach((sectionID) => {
+          this.rowIdentities.push(Object.keys(this._dataBlob[sectionID]));
+        });
+      }
+      this._cachedRowCount = countRows(this.rowIdentities);
+
+      // Dirty all rows and sections
+      this._calculateDirtyArrays([], [], []);
```

Comment:
TODO: have the constructor accept `prevData` param, so we don't do extra work when cloning